### PR TITLE
Add flag to generate a full, no-reflective implementation of messages.

### LINKED ISSUE #412: Reflection-free code generation.
* [ ] Non-reflecton based adapter on messages
* [ ] Enum adapters
* [ ] Generated `toString`

Triggered by JavaGenerator options.

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #412.

### CODE REVIEW

File: wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java [hunk 1]
```diff
@@ -90,7 +93,8 @@
           : Collections.<String>emptyList();
       JavaGenerator javaGenerator = JavaGenerator.get(schema)
           .withOptions(!noOptions, enumOptionsList)
-          .withAndroid(emitAndroid);
+          .withAndroid(emitAndroid)
+          .withFull(emitFull);
```
Comment (JakeWharton) [inline wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java]: I don't like this API. I think we want a proper builder.

File: wire-compiler/src/main/java/com/squareup/wire/CommandLineOptions.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
   public static final String QUIET_FLAG = "--quiet";
   public static final String DRY_RUN_FLAG = "--dry_run";
   public static final String ANDROID = "--android";
+  public static final String FULL = "--full";
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/CommandLineOptions.java]: don't love this name

Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/CommandLineOptions.java]: I had `--no_reflection` initially

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 3]
```diff
@@ -504,6 +545,193 @@ public TypeSpec generateMessage(MessageType type) {
...
+    result.beginControlFlow("switch (tag)");
+
+    for (Field field : type.fieldsAndOneOfFields()) {
+      int fieldTag = field.tag();
+      String fieldName = nameAllocator.get(field);
+      CodeBlock adapter = singleAdapterFor(field);
+
+      if (isEnum(field.type())) {
+        result.beginControlFlow("case $L:", fieldTag);
+        result.beginControlFlow("try");
```
Comment (swankjesse) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: Ooh I think this is going to add braces. That's okay, but a bit surprising

File: wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java (hunk 1 above)
Comment (swankjesse) [inline wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java]: Yep.

**Comment by swankjesse** [comment]
# LGTM

**Pull request merged**